contracts/finance/ProtocolStaking.sol

```diff
@@ -0,0 +1,286 @@
+import {SafeCast} from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+import {Checkpoints} from "@openzeppelin/contracts/utils/structs/Checkpoints.sol";
+import {Time} from "@openzeppelin/contracts/utils/types/Time.sol";
+
+interface IERC20Mintable is IERC20 {
+    function mint(address to, uint256 amount) external;
+}
+
+contract ProtocolStaking is AccessControlDefaultAdminRulesUpgradeable, ERC20VotesUpgradeable, UUPSUpgradeable {
+    using Checkpoints for Checkpoints.Trace208;
+    using SafeERC20 for IERC20;
+    using Math for uint256;
+
+    /// @custom:storage-location erc7201:zama.storage.ProtocolStaking
+    struct ProtocolStakingStorage {
+        // Stake - general
+        address _stakingToken;
+        uint256 _totalEligibleStakedWeight;
+        // Stake - release
+        uint256 _unstakeCooldownPeriod;
+        mapping(address => Checkpoints.Trace208) _unstakeRequests;
+        mapping(address => uint256) _released;
+        // Reward - issuance curve
+        uint256 _lastUpdateTimestamp;
+        uint256 _lastUpdateReward;
+        uint256 _rewardRate;
+        // Reward - recipient
+        mapping(address => address) _rewardsRecipient;
+        // Reward - payment tracking
+        mapping(address => int256) _paid;
+        int256 _totalVirtualPaid;
+    }
+
+    // keccak256(abi.encode(uint256(keccak256("zama.storage.ProtocolStaking")) - 1)) & ~bytes32(uint256(0xff))
+    bytes32 private constant PROTOCOL_STAKING_STORAGE_LOCATION =
+        0x6867237db38693700f305f18dff1dbf600e282237f7d452b4c792e6b019c6b00;
+    bytes32 private constant ELIGIBLE_ACCOUNT_ROLE = keccak256("eligible-account-role");
+
+    event TokensStaked(address indexed account, uint256 amount);
+    event TokensUnstaked(address indexed account, address indexed recipient, uint256 amount);
+    event RewardRateSet(uint256 rewardRate);
+    event UnstakeCooldownPeriodSet(uint256 unstakeCooldownPeriod);
+    event RewardsRecipientSet(address indexed account, address indexed recipient);
+
+    error InvalidAmount();
+    error EligibleAccountAlreadyExists(address account);
+    error EligibleAccountDoesNotExist(address account);
+    error TransferDisabled();
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(
+        string memory name,
+        string memory symbol,
+        string memory version,
+        address stakingToken_,
+        address governor
+    ) public virtual initializer {
+        __AccessControlDefaultAdminRules_init(0, governor);
+        __ERC20_init(name, symbol);
+        __EIP712_init(name, version);
+        _getProtocolStakingStorage()._stakingToken = stakingToken_;
+    }
+
+    /// @dev Stake `amount` tokens from `msg.sender`.
+    function stake(uint256 amount) public virtual {
+        _mint(msg.sender, amount);
+        IERC20(stakingToken()).safeTransferFrom(msg.sender, address(this), amount);
```

Comment:
It looks safer to mint shares only after the funds are received in case of re-entrancy:
```suggestion
        IERC20(stakingToken()).safeTransferFrom(msg.sender, address(this), amount);
        _mint(msg.sender, amount);
```